src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs

```diff
@@ -422,9 +437,14 @@ private void AddLexedTokenSlot()
             }
             else
             {
-                var tmp = new ArrayElement<SyntaxToken>[_lexedTokens.Length * 2];
-                Array.Copy(_lexedTokens, tmp, _lexedTokens.Length);
-                _lexedTokens = tmp;
+                var lexedTokens = _lexedTokens;
+                if (lexedTokens.Length == CachedTokenArraySize)
+                {
+                    Array.Clear(lexedTokens, 0, lexedTokens.Length);
+                    s_lexedTokensPool.Free(lexedTokens);
+                }
+
+                Array.Resize(ref _lexedTokens, _lexedTokens.Length * 2);
```

Comment:
What am I missing in this code? Would this not clear `_lexedTokens`, `Free` it, and then resize the empty array, throwing away all the tokens that have been lexed so far?

Reply:
I did debug my changes, but I wasn't able to hit the else clauses here (and the if condition was complicated enough that I didn't want to mess with it)